Sources/System/Sockets/SocketDescriptor.swift

```diff
@@ -0,0 +1,259 @@
+/*
+ This source file is part of the Swift System open source project
+
+ Copyright (c) 2021 Apple Inc. and the Swift System project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+*/
+
+
+// TODO: @available(...)
+
+// TODO: Windows uses a SOCKET type; doesn't have file descriptor
+// equivalence
+
+/// TODO
+@frozen
+public struct SocketDescriptor: RawRepresentable, Hashable {
+  /// The raw C socket
+  @_alwaysEmitIntoClient
+  public let rawValue: CInt
+
+  /// Creates a strongly-typed socket from a raw C socket
+  @_alwaysEmitIntoClient
+  public init(rawValue: CInt) { self.rawValue = rawValue }
+}
+
+extension SocketDescriptor {
+  /// The file descriptor for `self`.
+  @_alwaysEmitIntoClient
+  public var fileDescriptor: FileDescriptor {
+    FileDescriptor(rawValue: rawValue)
+  }
+
+  /// Treat `fd` as a socket descriptor, without checking with the operating
+  /// system that it actually refers to a socket
+  @_alwaysEmitIntoClient
+  public init(unchecked fd: FileDescriptor) {
```

Comment:
I wonder if `unchecked` is the right argument label to use here. It makes me think of `Range(uncheckedBounds:)`, in which case the checking is a bounds check precondition.

Reply:
For `Range`, is that precondition later checked? Here, the failure mode is "recoverable", i.e. Errno is thrown, but it is still an operation that doesn't check that the result is valid until use. "unsafe" is flat out as this will be checked and does not produce UB.

Reply:
No -- the `unchecked` in `Range(uncheckedBounds:)` indicates that failing to fulfill the unchecked precondition will lead to undefined behavior.

This is not the case here at all -- the difference between FileDescriptor and SocketDescriptor is a figment of our own imagination; mixing up the two will do no harm.